### ISSUE TITLE
[leo_object_storage] failed to rollback the compaction status to idle from running

### ISSUE THREAD

**Issue by mocchira** [leo-project/leofs#511]
The discussion is here.
https://groups.google.com/forum/#!topic/leoproject_leofs/UZtQHe2YjMU

The below sentence in this discussion is the focus on this issue.
> As for the reason why this inconsistency happened, 
> there should be a case we have been missing to rollback the compaction 
> status into **idle** from **running** when something goes wrong 
> so we will look into that right away.

**Comment by mocchira** [comment]
Errors happened while running diagnose at Vladimir's env.
- https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_haystack.erl#L844
- https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_haystack.erl#L722
  - https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_transformer.erl#L289
- https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_haystack.erl#L741

Where an abort can happen (all cases caused by an error returned by file:pread)
- https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_haystack.erl#L710
- https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_haystack.erl#L762
- https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_object_storage_haystack.erl#L833

**Comment by mocchira** [comment]
WIP

If leo_compact_fsm_worker:run is called with a timeout then skipping a garbage block may take much more time according to its garbage size, the compaction status may get inconsistent and diagnose|compaction stop unexpectedly in case of skipping a huge garbage block ( cause https://github.com/leo-project/leo_object_storage/blob/1.3.0/src/leo_compact_fsm_worker.erl#L850-L866 code block to be executed many times )

**Comment by mocchira** [comment]
#520 maybe related to this issue.

**Comment by mocchira** [comment]
Succeeded in reproducing this issue by adding a very large garbage block into a avs and
observed what happened in leo_storage.

```erlang
(storage_0@127.0.0.1)23> recon_trace:calls({leo_compact_fsm_worker, '_', fun(_) -> return_trace() end}, 10, [{scope, local}]).

2:18:25.103820 <0.181.0> leo_compact_fsm_worker:execute_1(ok, {compaction_worker_state,leo_compact_worker_0,leo_object_storage_0,
    leo_object_storage_read_0,leo_metadata_0,
    {backend_info,undefined,<<"LeoFS AVS-2.4">>,<<"LeoFS AVS-2.4">>,
        "/home/ubuntu/dev/leofs/package/leo_storage/avs/object/0.avs",
        "/home/ubuntu/dev/leofs/package/leo_storage/avs/object/0.avs_63646478525",
        {file_descriptor,prim_file,{#Port<0.4915>,384}},
        {file_descriptor,prim_file,{#Port<0.4921>,387}}},
    <0.1414.0>,leo_diagnosis_log_0,running,false,false,false,false,500,3000,0,
    1000,1500,10,
    {compaction_prms,<<>>,<<>>,
        {metadata_1,<<>>,0,0,0,0,0,0,0,0,0,0,0,0,undefined,0,0,0},
        143943606,209715394,
        #Fun<leo_redundant_manager_api.has_charge_of_node.2>,0,0,0,0},
    63646478525,194,
    {set,3,16,16,8,80,48,
        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
        {{[],[],
          [{invalid_format,unexpected_time_format}],
          [],[],[],
          [{845,"invalid data"}],
          [],[],[],[],[],[],
          [{776,"data-size did not match"}],
          [],[]}}},
    [],undefined}, 1)

## show the same function(execute_1) called repeatedly while increasing the offset
## the throughput is terribly slow around 100MB/hour...

Recon tracer rate limit tripped.
```

so that means we need to fix #520 .
and there is another observation.

```erlang
(storage_0@127.0.0.1)23> sys:get_state(leo_compact_fsm_controller).
{running,{state,undefined,
                [{leo_compact_worker_0,leo_object_storage_0},
                 {leo_compact_worker_1,leo_object_storage_1},
                 {leo_compact_worker_2,leo_object_storage_2},
                 {leo_compact_worker_3,leo_object_storage_3},
                 {leo_compact_worker_4,leo_object_storage_4},
                 {leo_compact_worker_5,leo_object_storage_5},
                 {leo_compact_worker_6,leo_object_storage_6},
                 {leo_compact_worker_7,leo_object_storage_7}],
                [{<0.5001.0>,leo_object_storage_0},
                 {<0.1416.0>,leo_object_storage_2},
                 {<0.1415.0>,leo_object_storage_1},
                 {<0.1414.0>,leo_object_storage_0}],
                1,true,false,undefined,8,[],
                [leo_object_storage_2,leo_object_storage_3,
                 leo_object_storage_4,leo_object_storage_5,
                 leo_object_storage_6,leo_object_storage_7],
                [leo_object_storage_1],
                [],
                [{<0.5001.0>,true}],
                63646478828,
                [undefined],
                running}}
```

pid_pairs in the state has gotten an unexpected state ( there should be only {<0.5001.0>, leo_object_storage_0} ) so that means there is a bug in leo_compact_fsm_controller.
